misk/src/main/kotlin/misk/clustering/DefaultCluster.kt

```diff
@@ -0,0 +1,156 @@
+      { ClusterHashRing(it) }
+) : AbstractExecutionThreadService(), Cluster, DependentService {
+  private val snapshotRef = AtomicReference<Cluster.Snapshot>(Cluster.Snapshot(
+      self = self,
+      selfReady = false,
+      readyMembers = setOf(),
+      resourceMapper = newResourceMapper(setOf())
+  ))
+  private val running = AtomicBoolean(false)
+  private val actions = BlockingArrayQueue<(MutableSet<ClusterWatch>) -> Unit>()
+
+  override val snapshot: Cluster.Snapshot get() = snapshotRef.get()
+  override val consumedKeys: Set<Key<*>> = setOf()
+  override val producedKeys: Set<Key<*>> = setOf(keyOf<Cluster>())
+
+  /**
+   * Runs the internal event loop that handles requests to add watches or cluster changes. We use
+   * a single threaded event loop to ensure that watches get consistent diffs for the cluster
+   * without being racy or requiring locks that span application code (which might result in
+   * deadlocks). When a watch is registered, it is provided with the cluster membership as it
+   * is known at that time, and then receives diffs as they arrive.
+   */
+  override fun run() {
+    val watches = mutableSetOf<ClusterWatch>()
+
+    while (running.get()) {
+      try {
+        val action = actions.take()
+        action(watches)
+      } catch (th: Throwable) {
+        log.error(th) { "error calling event queue action" }
+      }
+    }
+  }
+
+  public override fun startUp() {
+    log.info { "starting cluster manager" }
+    running.set(true)
+  }
+
+  public override fun shutDown() {
+    log.info { "shutting down cluster manager" }
+    running.set(false)
+    actions.add { _ -> } // empty action to up the action queue if it is empty
+  }
+
+  /** Adds a new cluster watch */
+  override fun watch(watch: ClusterWatch) {
+    actions.add { watches ->
+      watches.add(watch)
+      watch(Cluster.Changes(snapshot))
+    }
+  }
+
+  /** Triggers a change to the cluster in response members becoming ready or not ready */
+  fun clusterChanged(
+    membersBecomingReady: Set<Cluster.Member> = setOf(),
+    membersBecomingNotReady: Set<Cluster.Member> = setOf()
+  ) {
+    actions.add { watches ->
+      val changeTracker = ClusterChangeTracker(this)
+      membersBecomingReady.forEach { changeTracker.memberReady(it) }
+      membersBecomingNotReady.forEach { changeTracker.memberNotReady(it) }
+      val changes = changeTracker.computeChanges()
+
+      if (changes.hasDiffs) {
+        snapshotRef.set(changes.snapshot)
+        watches.forEach {
+          try {
+            it(changes)
+          } catch (th: Throwable) {
+            log.error(th) { "error triggering watch in response to cluster change" }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Triggers a callback once all of the actions on the queue have been processed. Useful
+   * for writing deterministic tests
+   */
+  internal fun syncPoint(callback: () -> Unit) {
```

Comment:
instead of taking a callback should this just create its own CountDownLatch?

Reply:
I think I prefer leaving it open ended for the tests, but don't feel strongly about it